Move all of DataProtection and Extensions.Identity to netstandard2.0

HaoK opened the pull request:
First part of https://github.com/aspnet/AspNetCore/issues/3774 this will allow Extensions.Identity.Core Stores, and a few parts of data protection to target netstandard 2.0

HaoK committed:
Revert "Use CryptographicOperations.FixedTimeEquals (#9798)"

This reverts commit 2b2aeee9f869ec43f65c188a66d9285858823628.

HaoK committed:
Revert "Use CryptographicOperations.FixedTimeEquals (#9798)"

This reverts commit 2b2aeee9f869ec43f65c188a66d9285858823628.

HaoK committed:
Switch some of DP and Identity.Core to netstandard 2.0

HaoK committed:
Switch some of DP and Identity.Core to netstandard 2.0

HaoK committed:
Move stores to netstandard2.0

HaoK committed:
Move stores to netstandard2.0

HaoK commented:
@natemcmaster @blowdart this seems a bit scary as the NetCorePbkdf2Provider provider is using a netcore DeriveBytes API, I only started with a few of the DP.abstractions packages and KeyDerivation which is the main thing that Identity needs for the password hasher.  

I also reverted the Apr 26 PR that switched the password hasher to use CryptographicOperations.FixedTimeEquals  (2b2aeee9f869ec43f65c188a66d9285858823628)

natemcmaster commented on src/DataProtection/Cryptography.KeyDerivation/src/Microsoft.AspNetCore.Cryptography.KeyDerivation.csproj:
```diff
@@ -2,7 +2,7 @@
   <PropertyGroup>
     <Description>ASP.NET Core utilities for key derivation.</Description>
-    <TargetFramework>netcoreapp3.0</TargetFramework>
+    <TargetFramework>netstandard2.0</TargetFramework>
```
```suggestion
    <TargetFrameworks>netstandard2.0;netcoreapp3.0</TargetFrameworks>
```

natemcmaster replied:
On second thought, this should probably stay the same as it was in 2.x and target `netstandard2.0;netcoreapp2.0`. By changing the TFMs to `netstandard2.0;netcoreapp3.0`, anyone who uses the 3.0 package on netcoreapp2.x will end up with different behavior because they will get the netstandard DLL, not the netcore DLL. This will have a negative perf impact on Linux/macOS.

cref https://github.com/aspnet/AspNetCore/commit/e552b5861a9f35a3daa3079d0e56b011f907daed

natemcmaster commented on src/DataProtection/Cryptography.KeyDerivation/src/PBKDF2/NetCorePbkdf2Provider.cs:
```diff
@@ -1,63 +1,63 @@
-using System.Diagnostics;
-using System.Security.Cryptography;
-using System.Text;
+//using System;
```
Undo this and use `#if` to only compile this for netcoreapp.

natemcmaster commented on src/Identity/Extensions.Core/src/PasswordHasher.cs:
```diff
@@ -222,7 +242,7 @@ private static bool VerifyHashedPasswordV2(byte[] hashedPassword, string passwor
             // Hash the incoming password and verify it
             byte[] actualSubkey = KeyDerivation.Pbkdf2(password, salt, Pbkdf2Prf, Pbkdf2IterCount, Pbkdf2SubkeyLength);
-            return CryptographicOperations.FixedTimeEquals(actualSubkey, expectedSubkey);
+            return ByteArraysEqual(actualSubkey, expectedSubkey);
```
```suggestion
#if NETSTANDARD2_0
            return ByteArraysEqual(actualSubkey, expectedSubkey);
#elif NETCOREAPP3_0
            return CryptographicOperations.FixedTimeEquals(actualSubkey, expectedSubkey);
#else
#error Update target frameworks
#endif
```

Let's not undo https://github.com/aspnet/AspNetCore/pull/9798

natemcmaster commented on src/Identity/Extensions.Core/src/Microsoft.Extensions.Identity.Core.csproj:
```diff
@@ -2,7 +2,7 @@
   <PropertyGroup>
     <Description>ASP.NET Core Identity is the membership system for building ASP.NET Core web applications, including membership, login, and user data. ASP.NET Core Identity allows you to add login features to your application and makes it easy to customize data about the logged in user.</Description>
-    <TargetFramework>netcoreapp3.0</TargetFramework>
+    <TargetFramework>netstandard2.0</TargetFramework>
```
```suggestion
    <TargetFrameworks>netstandard2.0;netcoreapp3.0</TargetFrameworks>
```

natemcmaster commented on src/Identity/Extensions.Core/src/PasswordHasher.cs:
```diff
@@ -65,6 +65,26 @@ public PasswordHasher(IOptions<PasswordHasherOptions> optionsAccessor = null)
             _rng = options.Rng;
         }
 
+        // Compares two byte arrays for equality. The method is specifically written so that the loop is not optimized.
```
Would recommend wrapping this in `#if NETSTANDARD2_0` to avoid usage in netcoreapp3.0, where there is a better API.

natemcmaster reviewed (changes requested):
Good start, but let's use cross compiling to keep existing improvements.

HaoK committed:
Update src/DataProtection/Cryptography.KeyDerivation/src/Microsoft.AspNetCore.Cryptography.KeyDerivation.csproj

Co-Authored-By: Nate McMaster <natemcmaster@users.noreply.github.com>

HaoK committed:
Update src/DataProtection/Cryptography.KeyDerivation/src/Microsoft.AspNetCore.Cryptography.KeyDerivation.csproj

Co-Authored-By: Nate McMaster <natemcmaster@users.noreply.github.com>

HaoK committed:
Update src/Identity/Extensions.Core/src/PasswordHasher.cs

Co-Authored-By: Nate McMaster <natemcmaster@users.noreply.github.com>

HaoK committed:
Update src/Identity/Extensions.Core/src/PasswordHasher.cs

Co-Authored-By: Nate McMaster <natemcmaster@users.noreply.github.com>

HaoK committed:
Update src/Identity/Extensions.Core/src/Microsoft.Extensions.Identity.Core.csproj

Co-Authored-By: Nate McMaster <natemcmaster@users.noreply.github.com>

HaoK committed:
Update src/Identity/Extensions.Core/src/Microsoft.Extensions.Identity.Core.csproj

Co-Authored-By: Nate McMaster <natemcmaster@users.noreply.github.com>

HaoK committed:
PR feedback

HaoK committed:
PR feedback

HaoK committed:
Update a bunch of refs

HaoK committed:
Update a bunch of refs

HaoK committed:
Put dataprotection back to netcoreapp3.0 for now so things build

HaoK committed:
Put dataprotection back to netcoreapp3.0 for now so things build

HaoK commented:
Okay so I took a stab at switching DP to netstandard and the issue is the IStartupFilter that gets added as part of `AddDataProtection()`

```
           services.TryAddEnumerable(ServiceDescriptor.Singleton<IStartupFilter, DataProtectionStartupFilter>());
```

Was there any prior discussion about how this should be resolved?

HaoK committed:
Update refs

HaoK committed:
Update refs

HaoK commented:
@natemcmaster @ajcvickers what do you guys think about merging the identity + some of dataprotection changes since this at least enables extensions identity + dp abstractions / keyderivations to be used on netstandard 2.0.  I can open another PR addressing the gnarly hosting + DP when we have agreed on an approach

natemcmaster commented:
Discussed briefly with @ajcvickers.

I'm okay merging the changes in this PR as long as you update the description to remove the "fixes #3774" description. The goal of https://github.com/aspnet/AspNetCore/issues/3774 was to support packages which are both .NET Standard _and_ do not have a dependency on the .NET Core shared framework, and this PR doesn't yet satisfy that criteria. To complete that issue, you'll need to address the DP -> Hosting dependency. Here are my thoughts on ways you might solve this: https://github.com/aspnet/AspNetCore/issues/3774#issuecomment-500563552.

Also, to make the packages available as packages, you need to set `<IsShippingPackage>` in the .csproj file, otherwise the assemblies will still be shared-framework only.

HaoK committed:
Update Microsoft.AspNetCore.DataProtection.Abstractions.csproj

HaoK committed:
Update Microsoft.AspNetCore.DataProtection.Abstractions.csproj

HaoK committed:
Update Microsoft.AspNetCore.Cryptography.Internal.csproj

HaoK committed:
Update Microsoft.AspNetCore.Cryptography.Internal.csproj

HaoK committed:
Update Microsoft.AspNetCore.Cryptography.KeyDerivation.csproj

HaoK committed:
Update Microsoft.AspNetCore.Cryptography.KeyDerivation.csproj

HaoK committed:
Update Microsoft.Extensions.Identity.Core.csproj

HaoK committed:
Update Microsoft.Extensions.Identity.Core.csproj

HaoK committed:
Update Microsoft.Extensions.Identity.Stores.csproj

HaoK committed:
Update Microsoft.Extensions.Identity.Stores.csproj

HaoK commented:
Removed fix comment and added `<IsShippingPackage>true</IsShippingPackage`

HaoK committed:
Update Microsoft.AspNetCore.DataProtection.Abstractions.csproj

HaoK committed:
Update Microsoft.AspNetCore.DataProtection.Abstractions.csproj

HaoK committed:
Update Microsoft.AspNetCore.DataProtection.Abstractions.csproj

HaoK committed:
Update Microsoft.AspNetCore.DataProtection.Abstractions.csproj

dougbu committed:
Do not exit cleanly after failed `msbuild` in `build.ps1`'s `finally` block
- part of aspnet/AspNetCore-Internal#2665
- should surface errors we're currently hiding

Also increase PowerShell logging to help debug build and test hangs

nit: Run `finally` block after catching an exception

HaoK committed:
Switch Startup filter to HostedService

HaoK committed:
Switch Startup filter to HostedService

HaoK committed:
Update refs

HaoK committed:
Update refs

HaoK committed:
Remove netcoreapp ref

HaoK committed:
Remove netcoreapp ref

HaoK committed:
Update all the rest of the packages to netstandard2/2.1

HaoK committed:
Update all the rest of the packages to netstandard2/2.1

HaoK commented on src/Shared/WebEncoders/WebEncoders.cs:
```diff
@@ -2,7 +2,9 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+#if NETCOREAPP3_0
```
FYi @BrennanConroy I had to add this guard around the using buffers introduced from the spanify web encoders change last week since it isn't available in netstandard 2.0

BrennanConroy replied:
Makes sense 👍

HaoK commented:
This should now fix https://github.com/aspnet/AspNetCore/issues/3774 pending the resolution of IHostedService not running before the server starts :(

dougbu commented on src/DataProtection/DataProtection/src/Microsoft.AspNetCore.DataProtection.csproj:
```diff
@@ -2,7 +2,7 @@
     <Description>ASP.NET Core logic to protect and unprotect data, similar to DPAPI.</Description>
-    <TargetFramework>netcoreapp3.0</TargetFramework>
+    <TargetFramework>netstandard2.0</TargetFramework>
     <IsAspNetCoreApp>true</IsAspNetCoreApp>
```
Suggest we also need to ship this as a package i.e.
```suggestion
    <IsAspNetCoreApp>true</IsAspNetCoreApp>
    <IsShippingPackage>true</IsShippingPackage>
```

HaoK committed:
Update src/DataProtection/DataProtection/src/Microsoft.AspNetCore.DataProtection.csproj

Co-Authored-By: Doug Bunting <6431421+dougbu@users.noreply.github.com>

HaoK committed:
Update src/DataProtection/DataProtection/src/Microsoft.AspNetCore.DataProtection.csproj

Co-Authored-By: Doug Bunting <6431421+dougbu@users.noreply.github.com>

Tratcher commented:
> This should now fix #3774 pending the resolution of IHostedService not running before the server starts :(

In WebHost IHostedServices are started just after the server. In generic Host the web app is an IHostedService and they're started in registration order. Try testing with generic host to make sure you can get the ordering right. For WebHost it looks like IHostedServices is better than first-request, but not quite as good as IStartupFilter.

HaoK committed:
Update SharedFramework.Local.props

HaoK committed:
Update SharedFramework.Local.props

HaoK committed:
Update Microsoft.AspNetCore.DataProtection.Extensions.csproj

HaoK committed:
Update Microsoft.AspNetCore.DataProtection.Extensions.csproj

HaoK committed:
Update SharedFramework.Local.props

HaoK committed:
Update SharedFramework.Local.props

HaoK committed:
Add generic host key ring loading test

HaoK committed:
Add generic host key ring loading test

HaoK commented:
@Tratcher Cool, so the key ring startup test does pass using the generic host, but is this 'good enough'?  The webhost test fails, from @natemcmaster comment in the main issue, it doesn't seem safe for the key ring to not be intialized before server start, @blowdart time is running out for preview 7 do we need to resolve the key loading issue with WebHost now, or is that issue something we can tackle in a follow up PR (most likely early preview 8)?

HaoK committed:
Start hosted services before server

HaoK committed:
Start hosted services before server

HaoK commented:
@Tratcher I went ahead and tried starting IHostingService in WebHost before server start as well, I had to updated 3 tests as well, no idea if this is 'correct' but it makes the data protection web host test pass as well 🤷‍♂

Tratcher commented on src/Hosting/Hosting/src/Internal/WebHost.cs:
```diff
@@ -158,8 +162,6 @@ public virtual async Task StartAsync(CancellationToken cancellationToken = defau
             _applicationLifetime?.NotifyStarted();
 
-            // Fire IHostedService.Start
-            await _hostedServiceExecutor.StartAsync(cancellationToken).ConfigureAwait(false);
```
Can't change this, it's quite breaking.

HaoK replied:
The only reason I tried this was that @davidfowl seemed open to changing WebHost in https://github.com/aspnet/AspNetCore/issues/3774#issuecomment-501794264

> Another option might be an interface like IHostedService? We can change the WebHost to make sure these run before the server starts (I believe it doesn't today).

davidfowl replied:
Can't? We can discuss how breaking it is.

Tratcher commented on src/DataProtection/DataProtection/test/HostingTests.cs:
```diff
@@ -46,6 +47,32 @@ public async Task LoadsKeyRingBeforeServerStarts()
+                    .Services
+                    .Replace(ServiceDescriptor.Singleton(mockKeyRing.Object))
+                    .AddSingleton<IServer>(
+                        new FakeServer(onStart: () => tcs.TrySetException(new InvalidOperationException("Server was started before key ring was initialized")))));
```
This IServer isn't actually getting started, you didn't add the webhostedservice.

HaoK replied:
You mean adding a `.ConfigureWebHost(b => b.UseStartup<TestStartup>());` right?

Tratcher reviewed (commented):
As to if this is good enough for webhost, I honestly couldn't say. It's certainly not as good as the startup filter. I'll follow with you directly to bainstorm.

BrennanConroy committed:
Temporarily skip SignalR TS client tests (#11398)

mikaelm12 committed:
Make SSEMessageFormatter internal and use new InternalsVisibleTo pattern (#11356)

jkotalik committed:
Add file manifest for runtime and targeting pack (#11235)

jkotalik committed:
Remove copy race for IIS projects (#11394)

jkotalik committed:
Convert TLS connection adapter to connection middleware (#11109)

HaoK committed:
Add ConfigureWebHost to generic test

HaoK committed:
Add ConfigureWebHost to generic test

Commit:
Fix Template test runs on macOS and Linux (#11407)

SteveSandersonMS committed:
On RenderTreeBuilder, add SetUpdatesAttributeName (#11375)

HaoK committed:
Supressbaseline references

HaoK committed:
Supressbaseline references

mikaelm12 committed:
Add Invoke API with with no completion result (#11399)

BrennanConroy committed:
Pass options to HubConnectionContext (#11382)

HaoK committed:
Update Microsoft.Extensions.Identity.Stores.csproj

HaoK committed:
Update Microsoft.Extensions.Identity.Stores.csproj

ajaybhargavb committed:
Allow @bind-value to be specified on its own (#11401)

analogrelay committed:
Quarantine some tests (#11420)

davidfowl committed:
Remove connection adapters and move things to middleware (#11412)

* Remove connection adapters and move things to middleware
- Remove connection adapters from the public API surface (pubternal) and replace the existing adapters with connection middleware.
- Updated the tests

mikaelm12 committed:
Kestrel Override Client Cert Validation (#11423)

halter73 committed:
Work around potential race in PipeWriter (IIS Edition) (#11165)

Tratcher committed:
Remove redundant call to UseAuthentication (#11426)

Tratcher committed:
Avoid calling AuthenticateAsync for IIS out-of-proc #7750 (#11390)

Commit:
Use Enterprise Queue to reduce provisioning issues (#11434)

HaoK committed:
Update Microsoft.AspNetCore.DataProtection.csproj

dotnet-maestro[bot] committed:
[master] Update dependencies from aspnet/AspNetCore-Tooling aspnet/Blazor (#11307)

* Update dependencies from https://github.com/aspnet/AspNetCore-Tooling build 20190617.1

- Microsoft.NET.Sdk.Razor - 3.0.0-preview7.19317.1
- Microsoft.CodeAnalysis.Razor - 3.0.0-preview7.19317.1
- Microsoft.AspNetCore.Razor.Language - 3.0.0-preview7.19317.1
- Microsoft.AspNetCore.Mvc.Razor.Extensions - 3.0.0-preview7.19317.1

* Update dependencies from https://github.com/aspnet/Blazor build 20190617.1

- Microsoft.AspNetCore.Blazor.Mono - 0.10.0-preview7.19317.1

* Update dependencies from https://github.com/aspnet/AspNetCore-Tooling build 20190619.3

- Microsoft.NET.Sdk.Razor - 3.0.0-preview7.19319.3
- Microsoft.CodeAnalysis.Razor - 3.0.0-preview7.19319.3
- Microsoft.AspNetCore.Razor.Language - 3.0.0-preview7.19319.3
- Microsoft.AspNetCore.Mvc.Razor.Extensions - 3.0.0-preview7.19319.3

* Update dependencies from https://github.com/aspnet/AspNetCore-Tooling build 20190620.1

- Microsoft.NET.Sdk.Razor - 3.0.0-preview7.19320.1
- Microsoft.CodeAnalysis.Razor - 3.0.0-preview7.19320.1
- Microsoft.AspNetCore.Razor.Language - 3.0.0-preview7.19320.1
- Microsoft.AspNetCore.Mvc.Razor.Extensions - 3.0.0-preview7.19320.1

* Update dependencies from https://github.com/aspnet/AspNetCore-Tooling build 20190620.3

- Microsoft.NET.Sdk.Razor - 3.0.0-preview7.19320.3
- Microsoft.CodeAnalysis.Razor - 3.0.0-preview7.19320.3
- Microsoft.AspNetCore.Razor.Language - 3.0.0-preview7.19320.3
- Microsoft.AspNetCore.Mvc.Razor.Extensions - 3.0.0-preview7.19320.3

HaoK commented:
@aspnet/build any ideas how i can update SharedFramework.Local.props manually since I'm unable to run the script since my local restore/build is borked, it gave me a diff before when I updated it previously in this PR, but its not clear to me what the delta is between what it expects and what is checked in right now

```

eng\SharedFramework.Local.props(,): error : Generated code is not up to date in eng/SharedFramework.Local.props. You might need to regenerate the reference assemblies or project list (see docs/ReferenceAssemblies.md and docs/ReferenceResolution.md)
PowerShell exited with code '1'.
```

DylanDmitri committed:
Dylan/stack policy (#11293)

* Initial StackPolicy implementation

HaoK commented:
Figured out the diff was trailing whitespace in the sharedframework.local...grrr

HaoK committed:
Fix whitespace

HaoK committed:
Merge branch 'data-netstd' of https://github.com/aspnet/AspNetCore into data-netstd

HaoK commented:
Per offline thread with @Tratcher and @davidfowl this seems good enough for preview 7, will tweak in subsequent PRs as needed
